File: LEO/Domain/Recurrence/USFederalHolidays.swift
import Foundation

/// Determines whether a date falls on a US federal holiday (observed).
/// "Observed" rule: if the holiday falls on Saturday, observed the preceding Friday;
/// if it falls on Sunday, observed the following Monday.
public enum USFederalHolidays {

    public static func isHoliday(_ date: Date, cal: Calendar) -> Bool {
        let year = cal.component(.year, from: date)
        return holidays(for: year, cal: cal).contains { cal.isDate($0, inSameDayAs: date) }
    }

    /// All 11 federal holidays (observed dates) for a given year.
    public static func holidays(for year: Int, cal: Calendar) -> [Date] {
        var results: [Date] = []

        // New Year's Day — Jan 1
        results.append(observed(month: 1, day: 1, year: year, cal: cal))

        // Martin Luther King Jr. Day — 3rd Monday of January
        results.append(nthWeekday(.monday, n: 3, month: 1, year: year, cal: cal))

        // Presidents' Day — 3rd Monday of February
        results.append(nthWeekday(.monday, n: 3, month: 2, year: year, cal: cal))

        // Memorial Day — last Monday of May
        results.append(lastWeekday(.monday, month: 5, year: year, cal: cal))

        // Juneteenth — Jun 19
        results.append(observed(month: 6, day: 19, year: year, cal: cal))

        // Independence Day — Jul 4
        results.append(observed(month: 7, day: 4, year: year, cal: cal))

        // Labor Day — 1st Monday of September
        results.append(nthWeekday(.monday, n: 1, month: 9, year: year, cal: cal))

        // Columbus Day — 2nd Monday of October
        results.append(nthWeekday(.monday, n: 2, month: 10, year: year, cal: cal))

        // Veterans Day — Nov 11
        results.append(observed(month: 11, day: 11, year: year, cal: cal))

        // Thanksgiving — 4th Thursday of November
        results.append(nthWeekday(.thursday, n: 4, month: 11, year: year, cal: cal))

        // Christmas Day — Dec 25
        results.append(observed(month: 12, day: 25, year: year, cal: cal))

        return results.compactMap { $0 }
    }

    // MARK: - Private helpers

    /// Fixed-date holiday adjusted to the observed weekday.
    private static func observed(month: Int, day: Int, year: Int, cal: Calendar) -> Date {
        var comps = DateComponents(year: year, month: month, day: day)
        guard let date = cal.date(from: comps) else { return Date.distantPast }
        let weekday = cal.component(.weekday, from: date) // 1=Sun … 7=Sat
        if weekday == 7 { // Saturday → Friday
            return cal.date(byAdding: .day, value: -1, to: date) ?? date
        } else if weekday == 1 { // Sunday → Monday
            return cal.date(byAdding: .day, value: 1, to: date) ?? date
        }
        return date
    }

    /// Nth occurrence of a weekday within a month (1-based).
    private static func nthWeekday(_ weekday: Weekday, n: Int, month: Int, year: Int, cal: Calendar) -> Date {
        var comps = DateComponents(year: year, month: month, day: 1)
        guard let firstOfMonth = cal.date(from: comps) else { return Date.distantPast }
        let targetCalWD = weekday.calWeekday // 1=Sun … 7=Sat
        let firstWD = cal.component(.weekday, from: firstOfMonth)
        var offset = targetCalWD - firstWD
        if offset < 0 { offset += 7 }
        offset += (n - 1) * 7
        return cal.date(byAdding: .day, value: offset, to: firstOfMonth) ?? firstOfMonth
    }

    /// Last occurrence of a weekday within a month.
    private static func lastWeekday(_ weekday: Weekday, month: Int, year: Int, cal: Calendar) -> Date {
        // Find the 1st of next month, go back 1 day = last day of target month, then find the weekday
        var comps = DateComponents(year: year, month: month + 1, day: 1)
        if month == 12 { comps = DateComponents(year: year + 1, month: 1, day: 1) }
        guard let firstOfNext = cal.date(from: comps),
              let lastOfMonth = cal.date(byAdding: .day, value: -1, to: firstOfNext) else {
            return Date.distantPast
        }
        let targetCalWD = weekday.calWeekday
        let lastWD = cal.component(.weekday, from: lastOfMonth)
        var offset = targetCalWD - lastWD
        if offset > 0 { offset -= 7 }
        return cal.date(byAdding: .day, value: offset, to: lastOfMonth) ?? lastOfMonth
    }
}

File: apps/web/src/recurrence/usFederalHolidays.ts
/**
 * Port of LEO/Domain/Recurrence/USFederalHolidays.swift — a rule-based
 * computation (nth-weekday-of-month, last-weekday-of-month, Sat/Sun
 * observed-date shifting), not a hardcoded date table, so it's correct for
 * any year rather than needing manual upkeep. Used by the `skipUSHolidays`
 * LEORuleExtension filter in extensions.ts.
 *
 * "Observed" rule (matches Swift exactly): a holiday landing on Saturday is
 * observed the preceding Friday; on Sunday, the following Monday.
 */

function dateFrom(year: number, month: number, day: number): Date {
  // Local time, day-precision only — matches Calendar.current usage in the
  // Swift original; RRULE expansion elsewhere in this module also operates
  // in local time, so this stays consistent with the rest of the engine.
  return new Date(year, month - 1, day)
}

function isSameDay(a: Date, b: Date): boolean {
  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate()
}

/** Fixed-date holiday adjusted to the observed weekday — port of `observed()`. */
function observed(month: number, day: number, year: number): Date {
  const date = dateFrom(year, month, day)
  const weekday = date.getDay() // 0=Sun … 6=Sat
  if (weekday === 6) return new Date(date.getTime() - 86400_000) // Saturday -> Friday
  if (weekday === 0) return new Date(date.getTime() + 86400_000) // Sunday -> Monday
  return date
}

/** Nth occurrence of a weekday within a month (1-based) — port of `nthWeekday()`. */
function nthWeekday(weekday: number, n: number, month: number, year: number): Date {
  const firstOfMonth = dateFrom(year, month, 1)
  const firstWD = firstOfMonth.getDay()
  let offset = weekday - firstWD
  if (offset < 0) offset += 7
  offset += (n - 1) * 7
  return new Date(firstOfMonth.getTime() + offset * 86400_000)
}

/** Last occurrence of a weekday within a month — port of `lastWeekday()`. */
function lastWeekdayOfMonth(weekday: number, month: number, year: number): Date {
  const firstOfNext = month === 12 ? dateFrom(year + 1, 1, 1) : dateFrom(year, month + 1, 1)
  const lastOfMonth = new Date(firstOfNext.getTime() - 86400_000)
  const lastWD = lastOfMonth.getDay()
  let offset = weekday - lastWD
  if (offset > 0) offset -= 7
  return new Date(lastOfMonth.getTime() + offset * 86400_000)
}

const MONDAY = 1
const THURSDAY = 4

/** All 11 federal holidays (observed dates) for a given year — port of `holidays(for:)`. */
export function usFederalHolidays(year: number): Date[] {
  return [
    observed(1, 1, year), // New Year's Day
    nthWeekday(MONDAY, 3, 1, year), // MLK Day — 3rd Monday of January
    nthWeekday(MONDAY, 3, 2, year), // Presidents' Day — 3rd Monday of February
    lastWeekdayOfMonth(MONDAY, 5, year), // Memorial Day — last Monday of May
    observed(6, 19, year), // Juneteenth
    observed(7, 4, year), // Independence Day
    nthWeekday(MONDAY, 1, 9, year), // Labor Day — 1st Monday of September
    nthWeekday(MONDAY, 2, 10, year), // Columbus Day — 2nd Monday of October
    observed(11, 11, year), // Veterans Day
    nthWeekday(THURSDAY, 4, 11, year), // Thanksgiving — 4th Thursday of November
    observed(12, 25, year), // Christmas Day
  ]
}

export function isUSFederalHoliday(date: Date): boolean {
  return usFederalHolidays(date.getFullYear()).some((h) => isSameDay(h, date))
}
